### PR TITLE
OSD-14677: Added backplane-cli upgrade based on the latest version

### PR DESCRIPTION
**Pull request by samanthajayasinghe**
### Introduction
This PR introduces `ocm backplane upgrade` command that based on GitHub latest release 

### How to use it ?
```
$ ocm backplane upgrade
? A newer version "v0.0.1" is available.
Would you like to upgrade? (y/N)
```

### CODE REVIEW

File: internal/github/github.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	"golang.org/x/text/language"
+)
+
+var (
+	ErrRequestFailed   = errors.New("request failed")
+	ErrArchiveNotFound = errors.New("archive not found")
+)
+
+const (
+	gitHubApiEndPoint = "https://api.github.com/repos/samanthajayasinghe/backplane-cli" // temp repo
```
Comment (MitaliBhalla) [inline internal/github/github.go:26]: This needs a change, has to be the main repo link.

File: internal/github/github_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+			},
+			ExpectedVersion: "",
+			ExpectedAssets:  []upgrade.ReleaseAsset{},
+		},
+		"latest releases versions": {
+			Actual: upgrade.Release{
+				TagName: "v0.0.1",
+				Assets:  []upgrade.ReleaseAsset{},
+			},
+			ExpectedVersion: "v0.0.1",
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply internal/github/github_test.go:34]: It's `Actual.TagName` and mock via `httptest.NewServer()`, and there is no coupling with info.go on the tests

File: go.mod [hunk 3]
```diff
@@ -14,14 +14,25 @@ require (
...
 )
+
+require (
+	github.com/Masterminds/semver v1.5.0
+	github.com/inconshreveable/mousetrap v1.1.0 // indirect
+	github.com/pkg/errors v0.9.1
+	github.com/spf13/pflag v1.0.5 // indirect
+	golang.org/x/sys v0.4.0 // indirect
+	golang.org/x/text v0.6.0
+	gopkg.in/AlecAivazis/survey.v1 v1.8.8
```
Comment (supreeth7) [inline go.mod:37]: I'm skeptical about using third-party libraries, why do we need a lib for a simple command prompt?

**Comment by MitaliBhalla** [comment]
It will also be good to add comments in the code wherever you can.

File: internal/github/github_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+						DownloadUrl: "https://github.com/openshift/backplane-cli/releases/download/v0.0.1/backplane-cli_0.0.1_Linux_arm64.tar.gz",
+					},
+				},
+			},
+			osConfig: github.OSConfig{
+				OSType: "Linux",
+				OSArch: "arm64",
+			},
+			match:         true,
+			ExpectedAsert: "https://github.com/openshift/backplane-cli/releases/download/v0.0.1/backplane-cli_0.0.1_Linux_arm64.tar.gz",
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply internal/github/github_test.go:117]: It's actually the URL defined on L108 `LatestRelease.Assets[0].DownloadUrl`. We can change the URL but I feel it' s good to have actual data as a dummy URL

File: go.mod (hunk 3 above)
Comment (samanthajayasinghe) [reply go.mod:37]: survey.v1 lib : https://github.com/openshift/backplane-cli/blob/7d72524c95ccb96608c23b473f2368576fb7cbce/internal/upgrade/upgrade.go#L151

Comment (supreeth7) [reply go.mod:37]: Can't we do it the normal way?
ref: https://github.com/openshift/pagerduty-short-circuiter/blob/d7ed6a6796845680e2ee5a0b5ad424ffde46cc57/cmd/kite/teams/cmd.go#L93

**Comment by supreeth7** [comment]
Food for though: I think these changes are rather long and redundant in the long run, we could follow a similar approach to `osdctl`, [ref](https://github.com/openshift/osdctl/blob/master/pkg/utils/version.go) for this enablement.

**Comment by samanthajayasinghe** [comment]
> Food for thought: I think these changes are rather long and redundant in the long run, we could follow a similar approach to `osdctl`, [ref](https://github.com/openshift/osdctl/blob/master/pkg/utils/version.go) for this enablement.

Yes @supreeth7 , I follow the same approach to getting the latest release. In my case, I didn't hard code the asset URL but fetch it via the [API](https://api.github.com/repos/openshift/osdctl/releases/latest), 

Fetch the latest release: https://github.com/openshift/backplane-cli/blob/7d72524c95ccb96608c23b473f2368576fb7cbce/internal/github/github.go#L62

Fetch matching asset based on OS: https://github.com/openshift/backplane-cli/blob/7d72524c95ccb96608c23b473f2368576fb7cbce/internal/github/github.go#L171

File: internal/github/github_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (supreeth7) [reply internal/github/github_test.go:117]: Right as all these are test cases, covering them with hard-coded dummy/fake data is necessary here.

File: internal/github/github_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+						DownloadUrl: "https://github.com/openshift/backplane-cli/releases/download/v0.0.1/backplane-cli_0.0.1_Darwin_arm64.tar.gz",
+					},
+				},
+			},
+			osConfig: github.OSConfig{
+				OSType: "Darwin",
+				OSArch: "arm64",
+			},
+			match:         true,
+			ExpectedAsert: "https://github.com/openshift/backplane-cli/releases/download/v0.0.1/backplane-cli_0.0.1_Darwin_arm64.tar.gz",
```
[the remark this answers is not in the document]
Comment (supreeth7) [reply internal/github/github_test.go:100]: https://github.com/openshift/backplane-cli/pull/5#discussion_r1094350149

File: internal/github/github_test.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			match:         false,
+			ExpectedAsert: "",
+		},
+		"check matching asert for Mac": {
+			LatestRelease: upgrade.Release{
+				TagName: "v0.0.1",
+				Assets: []upgrade.ReleaseAsset{
+					{
+						Name:        "backplane-cli_0.0.1_Darwin_arm64.tar.gz",
+						DownloadUrl: "https://github.com/openshift/backplane-cli/releases/download/v0.0.1/backplane-cli_0.0.1_Darwin_arm64.tar.gz",
```
[the remark this answers is not in the document]
Comment (supreeth7) [reply internal/github/github_test.go:91]: https://github.com/openshift/backplane-cli/pull/5#discussion_r1094350149

File: go.mod (hunk 3 above)
Comment (supreeth7) [reply go.mod:37]: Can we do it without using any libs as it's just a user prompt? :)

File: cmd/ocm-backplane/upgrade/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+func long() string {
+	return strings.Join([]string{
+		"Upgrades the latest version release for",
+		"your machine's OS and architecture.",
+	}, " ")
+}
+
+var UpgradeCmd = &cobra.Command{
+	Use:   "upgrade",
+	Short: "Upgrades the current backplane-cli to latest version",
```
Comment (supreeth7) [inline cmd/ocm-backplane/upgrade/upgrade.go:23]:
```suggestion
	Short: "Upgrades the current backplane-cli to the latest version",
```

File: cmd/ocm-backplane/upgrade/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/openshift/backplane-cli/internal/github"
+	"github.com/openshift/backplane-cli/internal/upgrade"
+	"github.com/openshift/backplane-cli/pkg/info"
+	"github.com/spf13/cobra"
+)
+
+func long() string {
+	return strings.Join([]string{
+		"Upgrades the latest version release for",
+		"your machine's OS and architecture.",
```
Comment (supreeth7) [inline cmd/ocm-backplane/upgrade/upgrade.go:17]:
```suggestion
		"Upgrades the latest version release based on",
		"your machine's OS and architecture.",
```

**Comment by supreeth7** [comment]
Nice work @samanthajayasinghe, I really like the test coverage.
Please incorporate the suggestions and changes asked.

File: internal/github/github_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MitaliBhalla) [reply internal/github/github_test.go:117]: Ack, thanks both.

File: cmd/ocm-backplane/upgrade/upgrade.go (hunk 7 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/upgrade/upgrade.go:23]: Done

File: cmd/ocm-backplane/upgrade/upgrade.go (hunk 8 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/upgrade/upgrade.go:17]: Done

File: go.mod (hunk 3 above)
Comment (samanthajayasinghe) [reply go.mod]: All good .. Remove all unwanted dependencies on `upgrade` mod and use  `bufio.Reader` 
https://github.com/openshift/backplane-cli/blob/72c5601cb5a465492fb8b7853874b92e7b0d325a/internal/upgrade/upgrade.go#L149

**Comment by supreeth7** [comment]
I've tested this locally and can confirm this works elegantly, nice work!

Minor nit: Can you please display a message like this post upgrade?

> Backplane CLI has been upgraded to v0.0.1

**Comment by samanthajayasinghe** [comment]
> I've tested this locally and can confirm this works elegantly, nice work!
> 
> Minor nit: Can you please display a message like this post upgrade?
> 
> > Backplane CLI has been upgraded to v0.0.1

Thanks for the testing upgrade @supreeth7 
I added the post upgrade message

**Pull request merged**